hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/preloader/NearCachePreloader.java

```diff
@@ -0,0 +1,244 @@
+        byte[] sizeBytes = new byte[INT_SIZE_IN_BYTES];
+        int loadedKeys = 0;
+
+        Builder<Data> builder = InflatableSet.newBuilder(LOAD_BATCH_SIZE);
+        while (readFullyOrNothing(bis, sizeBytes)) {
+            int dataSize = toInt(sizeBytes);
+            byte[] payload = new byte[dataSize];
+            if (!readFullyOrNothing(bis, payload)) {
+                break;
+            }
+            builder.add(new HeapData(payload));
+            if (builder.size() == LOAD_BATCH_SIZE) {
+                adapter.getAll(builder.build());
+                builder = InflatableSet.newBuilder(LOAD_BATCH_SIZE);
+            }
+            loadedKeys++;
+        }
+        if (builder.size() > 0) {
+            adapter.getAll(builder.build());
+        }
+        return loadedKeys;
+    }
+
+    private void writeKeySet(FileOutputStream fos, FileChannel outChannel, Set<KS> keySet) throws IOException {
+        for (KS key : keySet) {
+            Data dataKey = serializationService.toData(key);
+            if (dataKey != null) {
+                int dataSize = dataKey.totalSize();
+
+                ensureBufHasRoom(fos, INT_SIZE_IN_BYTES);
+                buf.putInt(dataSize);
+
+                int position = 0;
+                int remaining = dataSize;
+                while (remaining > 0) {
+                    int transferredCount = Math.min(BUFFER_SIZE - buf.position(), remaining);
+                    ensureBufHasRoom(fos, transferredCount);
+                    buf.put(dataKey.toByteArray(), position, transferredCount);
+                    position += transferredCount;
+                    remaining -= transferredCount;
+                }
+
+                lastWrittenBytes += INT_SIZE_IN_BYTES + dataSize;
+                lastKeyCount++;
+            }
+            flushLocalBuffer(outChannel);
+        }
+    }
+
+    private void ensureBufHasRoom(FileOutputStream fos, int expectedSize) throws IOException {
+        if (buf.position() < BUFFER_SIZE - expectedSize) {
+            return;
+        }
+        fos.write(buf.array());
+        buf.position(0);
+    }
+
+    private void flushLocalBuffer(FileChannel outChannel) throws IOException {
+        if (buf.position() == 0) {
+            return;
+        }
+        buf.flip();
+        while (buf.hasRemaining()) {
+            outChannel.write(buf);
+        }
+        buf.clear();
+    }
+
+    private static String getStorageFileName(String configStorageFileName, String nearCacheName) {
+        if (!isNullOrEmpty(configStorageFileName)) {
+            return configStorageFileName;
+        }
+        return "nearcache-" + toFileName(nearCacheName) + ".store";
+    }
+
+    private static long getElapsedMillis(long started) {
+        return NANOSECONDS.toMillis(System.nanoTime() - started);
+    }
+
+    @SuppressWarnings({"checkstyle:magicnumber", "checkstyle:booleanexpressioncomplexity"})
+    private static int toInt(byte[] bytes) {
```

Comment:
I actually changed the write part as well to use the counterpart `Bits#writeIntB`, otherwise I depend on the `ByteBuffer` implementation if little or big endian is being used. I hope it's okay to stick with big endian for this, otherwise I have to re-create the binary storage files which are used in the load test.